### ISSUE TITLE
[Question] [TypeScript] Best way to decouple set() function

### ISSUE THREAD

**Issue by magiccookie** [pmndrs/zustand#628]
Hello! I wish to be able to dispatch a function instead of using setters and getters. But how is it better to deal with its type?

```typescript
interface MyStore {
  bears: number;
  honey: number;
  dispatch: (fn: HandlerFn | Partial<MyStore>) => void;
}

// Expected dispatch handler type to be something like:
type HandlerFn = (state: MyStore) => Partial<MyStore>;

export const useStore = create<MyStore>((set) => ({
  bears: 1,
  honey: 3,
  dispatch: (fn: HandlerFn | Partial<MyStore>) => set(fn)
}));

export const incBears = (state: MyStore) => ({
  bears: state.bears + 1,
  honey: state.honey + 3
});

export const incJustHoney = (state: MyStore) => ({
  honey: state.honey + 1
});

export const lotsOfBears = { bears: 999 };
```
example usage `<button onClick={() => dispatch(incBears)}>` . 

I expect my handler type to be as `HandlerFn` to return `Partial` store (or just `Partial` store in case of an object). 
However `Partial` doesn't match with the `set` interface (also I tried to compose anything nice out of zustand types)

The error is:
```
(parameter) fn: HandlerFn
Argument of type 'HandlerFn' is not assignable to parameter of type 'PartialState<MyStore, keyof MyStore, keyof MyStore, keyof MyStore, keyof MyStore>'.
  Type 'HandlerFn' is not assignable to type '(state: MyStore) => MyStore | Pick<MyStore, keyof MyStore>'.
    Type 'Partial<MyStore>' is not assignable to type 'MyStore | Pick<MyStore, keyof MyStore>'.
      Type 'Partial<MyStore>' is not assignable to type 'Pick<MyStore, keyof MyStore>'.
        Types of property 'bears' are incompatible.
          Type 'number | undefined' is not assignable to type 'number'.
            Type 'undefined' is not assignable to type 'number'.ts(2345)
```

Basically, if I just want to proxy the `set` function,  what would be the best way to annotate handler function?

I think I could have specified the handler types as for example `(state: MyStore) => Pick<MyStore, "bears", "honey">`, however, I don't always know which keys my function return

See full example here https://codesandbox.io/s/hopeful-cohen-xy33j?file=/src/store.ts

**Comment by dai-shi** [comment]
(pinning @TkDodo to share this use case.)

I'm not sure if there's a good way.
I would probably give up and do `set(fn as any)`. 😎 

If anyone has suggestions, please jump in.


----
As a side note, assuming we don't need to reuse `incBears` for multiple stores.
Here's what I would do:

```ts
interface MyStore {
  bears: number;
  honey: number;
}

export const useStore = create<MyStore>(() => ({
  bears: 1,
  honey: 3,
}));

export const dispatchIncBears = () => useStore.setState((state) => ({
  bears: state.bears + 1,
  honey: state.honey + 3
}));

export const dispatchIncJustHoney = () => useStore.setState((state) => ({
  honey: state.honey + 1
}));

export const dispatchLotsOfBears = () => useStore.setState({ bears: 999 });
```

This is the pattern how we often do in [valtio](https://github.com/pmndrs/valtio) and it's good for code splitting.

**Comment by magiccookie** [comment]
@dai-shi Thanks a lot for sharing

**Comment by TkDodo** [comment]
> I expect my handler type to be as HandlerFn to return Partial 

The problem with the `Partial` type is that it doesn't differentiate between a key not being present or a key being present, but having an `undefined` value.

so if `incBears` were to return a `Partial<Store>`, you could do:

```
export const incBears = (state: MyStore) => ({
  bears: state.bears + 1,
  honey: undefined // 🚨
});
```

and TypeScript would be totally fine with that. Since the `set` function internally just spreads whatever you return over the current state, you would wind up with `honey: undefined` in your store, even though you only allow numbers.

This is especially troublesome if you have a property that _can_ be undefined, and later refactor it to be required. In those situations, TypeScript wouldn't show you all the places where you set that value to `undefined`.

The type we have in zustand tries to _infer_ the type from the keys of the returned object, and it's inspired by how class based setState in react works.

> I'm not sure if there's a good way.
I would probably give up and do set(fn as any). 😎

I also couldn't find a better way, sorry.

**Comment by dai-shi** [comment]
@devanshj This is another type issue. Would be nice if we can solve this too.

**Comment by devanshj** [comment]
Yep it's already [solved][1] with the #662.

[1]: https://www.typescriptlang.org/play?jsx=0#code/JYOwLgpgTgZghgYwgAgLIE8DKYD2UUDeAUMsgEYRxQDOAXMiAK4C2FUA3CcgBY4gTp6TVtE6kAJsGoAHOGATd6AChgh6ACTghxAG2gAxEMgA+yAApUwwODoA8GbHggA+AJTIAvM+QA3HMHFOAF8iIgB6MOQAUQAPaQgESHFkSRk5BR4tXWhkMHR43JxyFGocZggwblAAc2QdYABrCFoiPILNbT0oQ09kJWowOWa0LFx8dy9zS2s7BzGXTiIIOLwwZAQ+AeRGaghHfF6EfCH7UacAGmR9iFPr52clfoqJ7yViUgoqOmQARnOuXj8QTIADM-wkUlk8kUfVUGiyXR6pgsUCsNlu8zcnm8uzAKhAriIQVcrkWy2kq3WmzWoAQACFKDRev1BpB6HMnC8+u9il96AMhgA6T5MgDUv3BPD4An5rIggsBAmQ4pBRNJoXJlI2IC2tIAUjswOppehmQK2SNrly3gCTbKhYrTeKfmqyStUVSdWsdDgwNQAPIwBlfXoEXk0egATmjyCCi1AkFgiBQ11sABVkMtINpqMgAKogBogHAAdxA2CGzi4YeqFQrFq4pFISi5aa4l1x9eajabtgAcmssxAc8h-YKUWi7Gn7j2m30GFmu-6oHnpOIhlB6AOTH0WUN6GmuQPCXO5x3eIwdOIAEoQaQ6ZMAfhap9f8+3Q5HLex20LxbLs5vqQj7IPogoAMI2HYA5VkBr5bmAgGvu4n7iLmbZwaeIFkDgOB6FoSFvvQ8A6LshEvIRpB+AE7bINQjBkNQRzABQL6nmGSj1AMw7QMo5rDGmlzSPgPjADgOxdgeXLUeIrhsUB36TDJoSvpcYQAFTqcgAAC4h3vgCBDMkZj4bs2xmQABvRjHMRQADqwCVJgEB6IkeAWcgzABNkJZUCg6lhIRth5rBcFKHUUjZrxfT8fQeZCSJYkSXKcXSf44iUcgHYuQkYzPjFKXIIeP55pllwQAAjowNjUPlShwHFlxkKlP44XhlAgJlcmZc2aU0a+ISkJcekDFAOCCIBineMppAhK0+QoAWRaluWcqeFwOBkAAVrlixtCg4HHJA62kOmmYxNmaH5n+K1dpcmDnZduaptOoXNnRcoAJIgI5MzAAAXtFXbfb9NgA9A6b3W9WXINI3BaLgzAg5OXb5RhTYUa+mChPtVxfT9k7g1AZ2obmS3-qtQz3Y9w5XS99wncgABk87hbi9D0wA2gARJ2crcwAusggGXLWYAc-M6bODzYtdoLIsfU4EtOFLgGY2+6OkCezNFXA1SppgVZEHpCAPgc2pbAAJJbAxK7+wBVSU6CsHhnC42mesG94HjIGGnPW7b+AC-lD1zSbZsoBbaxHJQFqHbHEChCAcDlGkSCjr7XCah6uMTjMUuM6QfsANLIKAyBNOgOAwEVwcHpzxdCyEc3J6nsjp-omekNnay45BOhTt7PYZqTfSCuPDULj40CcwLXKgDAOTXsgIHL0IEDT1ARJAA

**Comment by devanshj** [comment]
> it doesn't differentiate between a key not being present or a key being present

Can be fixed by using a stricter `Partial` ie `type Partial<T> = { [K in keyof T]?: T[K] }` and with `--exactOptionalPropertyTypes` set ;) Here's [before][1] and [after][2].

Also isn't `HandlerFn` same as `SetState<MyStore>`? Unless I'm missing something? And this [does compile][3].

[1]: https://www.typescriptlang.org/play?jsx=0#code/MYewdgzgLgBAZiEMC8MA8AVAfACigLhgwBoYBXQgBQEMAnKAS2oBtMsBKFLGAbwF8AUAIQgcPGKDJgCMAAww+pcZOmEpAEwCmcBmE3qF7IA

[2]: https://www.typescriptlang.org/play?jsx=0&exactOptionalPropertyTypes=true#code/MYewdgzgLgBAZiEMC8MA8AVAfACigLhgwBoYBXQgeQAUBDAJygEtaAbTLAShSxgG8AvgCghUAJ4AHAKYwaDZmw4p+MANoBpGEzAwA1lLEg4RALoB+Qhg0mYwoQhA4+MUGTAEYABlulnr94RuACZScNpSQbacQA

[3]: https://www.typescriptlang.org/play?jsx=0#code/JYWwDg9gTgLgBAYygUwIY2QGjgbzgZWRnxnWTgF84AzKCEOAIgC8BXAZ1IDsATRgbgBQg4FwxRqqBOQCyATxLRyOQXDgAjNFHYAuOF1YhNUIWoAWELsjl6DR5CdVwewdmHQIzegBTUuewmJSDAAeeUUUAD4ASjgAXki4ADcIYB4hCmFkAA9IWERLTjgOZAjyOMQUMjCFGCVI7292IliEuG8VNU1UbT0ARkwnCysbOABmQbUXNw8vdr8AohJq8Lqo1sTmmF8uaMEKaOihQRy8+ARC+FEEACEtdnj2zjI9VaUN9s6Ne71njAA6braOAAajgAyGlmsv2CyH+w2soPG+yOWVy0HOlzg1wAUhwYAAJKFyR5NWGvWrveKJDqQkYwsjw4lIvoo46nDEFLhFAA2EBg7AA8tQ7j0HhU8EDdHAAJxyyhCIA

**Comment by TkDodo** [comment]
> Can be fixed by using a stricter Partial ie type Partial<T> = { [K in keyof T]?: T[K] } and with --exactOptionalPropertyTypes set 

isn't that the exact implementation of the build in type `Partial`?

```
type Partial<T> = {
    [P in keyof T]?: T[P];
};
```

it's true that the new compiler flag solves it, but `exactOptionalPropertyTypes` is an opt in setting and only available in TS4.4+. However, I would be totally fine with moving there and pointing people in the right direction, which would basically mean going back to `Partial` for the type definition. It seems that the workarounds we introduced with
- #317 

has caused a bunch of other issues in edge-cases

**Comment by devanshj** [comment]
> isn't that the exact implementation of the build in type Partial?

Nope, the built in has an extra `| undefined` as in it's `T[K] | undefined` instead of `T[K]`.

> going back to Partial for the type definition

Right, although it's still kinda undecided if we'll remove the existing workaround or not.

> has caused a bunch of other issues in edge-cases

Are there other issues you're aware of (apart from this one)?

**Comment by TkDodo** [comment]
> Nope, the built in has an extra | undefined as in it's T[K] | undefined instead of T[K].

what I posted above is what I've copied from `lib.es5.d.ts` 🤔 :

![Screenshot 2021-11-16 at 16 05 35](https://user-images.githubusercontent.com/1021430/142010581-cc54a472-02ff-44c1-b74d-feeb91dec657.png)


> Are there other issues you're aware of (apart from this one)?

closed ones, which we worked around with 4 overloads:

- https://github.com/pmndrs/zustand/issues/393

**Comment by devanshj** [comment]
Look like you're right, I just looked at the tooltip and assumed that's the definition haha.

![image](https://user-images.githubusercontent.com/30295578/142012790-90972d2c-71ac-4e34-81a8-5481f44dad2d.png).

> closed ones, which we worked around with 4 overloads:

What do you mean by 4 overloads though? Could you produce a snippet where the current types would fail?

**Comment by TkDodo** [comment]
initially, the "better Partial" type proposed by me was:

```
export type PartialState<T extends State, K extends keyof T = keyof T> =
  | (Pick<T, K> | T)
  | ((state: T) => Pick<T, K> | T)
```

which was then extended to:

```
export type PartialState<
  T extends State,
  K1 extends keyof T = keyof T,
  K2 extends keyof T = K1,
  K3 extends keyof T = K2,
  K4 extends keyof T = K3
> =
  | (Pick<T, K1> | Pick<T, K2> | Pick<T, K3> | Pick<T, K4> | T)
  | ((state: T) => Pick<T, K1> | Pick<T, K2> | Pick<T, K3> | Pick<T, K4> | T)
```

because of the issue linked above because the type `K` could not be properly inferred.

All in all, with the new compiler option available, we could just revert all of this, use the built-in `Partial` type again (like it was before #317) and educate users that if they want real type-safety, they need to use TS4.4 with `exactOptionalPropertyTypes`.

**Comment by devanshj** [comment]
Yeah I guess that would be a better way to go, the only problem is `exactOptionalPropertyTypes` isn't turn on when using `tsc --init` so very few users would have it on. But on the other hand these workarounds have it's own problems. Though I think let's go with `Partial<T>`.

**Comment by dai-shi** [comment]
I think I like `exactOptionalPropertyTypes`. I will try enabling it in zustand/jotai/valtio.
